src/opuntia.erl

```diff
@@ -132,27 +140,81 @@ calculate(#token_bucket_shaper{shape = {MaximumTokens, Rate},
+    %% This is now a float and so will all below be, to preserve best rounding errors possible
+    convert_time_unit(NativeTimeSinceLastUpdate, native, TimeUnit).
 
-    %% We penalise rounding up, the most important contract is that rate will never exceed that
-    %% requested, but the same way timeouts in Erlang promise not to arrive any time earlier but
-    %% don't promise at what time in the future they would arrive, nor we promise any upper bound
-    %% to the limits of the shaper delay.
-    RoundedDelayMs = ceil(MaybeDelayMs),
+%% Unbounded growth is calculated, with float precision, in the configured time units
+%%
+%% Note that it can be negative, if earlier we have penalised giving a larger `last_update' and now we
+%% update even before we have reach the point in time where the previous `last_update' was set
+%%
+%% If the growth was negative, that means that it has grown the debt instead
+-spec unbounded_bucket_diff(rate(), tokens(), float()) -> float().
+unbounded_bucket_diff(Rate, LastAvailableTokens, TimeDiffInUnits) ->
+    %% How much we might have recovered since last time
+    AvailableAtGrowthRate = Rate * TimeDiffInUnits,
+    %% Unbounded growth at rate since the last update
+    LastAvailableTokens + AvailableAtGrowthRate.
 
-    NewShaper = Shaper#token_bucket_shaper{available_tokens = TokensAvailable,
-                                           last_update = NativeNow + RoundedDelayMs,
-                                           debt = RoundedDelayMs - MaybeDelayMs},
-    {NewShaper, RoundedDelayMs}.
+%% This is the real growth considering the maximum bucket size.
+-spec exactly_available_now(bucket_size(), float()) -> float().
+exactly_available_now(MaximumTokens, UnboundedTokenGrowth) ->
+    %% Real recovery cannot grow higher than the actual rate in the window frame
+    ExactlyAvailableNow0 = min(MaximumTokens, UnboundedTokenGrowth),
+    %% But it can't be negative either which can happen if we were already in debt,
+    %% but this is a debt we will pay when we calculate the final punishment in final_state
+    max(+0.0, ExactlyAvailableNow0).
+
+%% We penalise rounding up, the most important contract is that rate will never exceed that
+%% requested, but the same way timeouts in Erlang promise not to arrive any time earlier but
+%% don't promise at what time in the future they would arrive, nor we promise any upper bound
+%% to the limits of the shaper delay.
+%%
+%% Two cases, either:
+%%   Punish is positive: even after paying the old debt you incurr a new debt again
+%%   Punish is negative: I overpenalised you last time, you get off now but with a future bill
+final_state(Shaper, TimeUnit, Punish, NativeNow) when Punish >= +0.0 ->
+    DelayMs = convert_time_unit(Punish, TimeUnit, millisecond),
+    RoundedDelayMs = ceil(DelayMs),
+    OverPenalisedNow = RoundedDelayMs - DelayMs,
+    OverUsedRateNowInUnits = convert_time_unit(OverPenalisedNow, millisecond, TimeUnit),
```

Comment:
Something is off with the naming for me, because `OverUsedRateNow...` would suggest using too many tokens, while it is over-penalised, not over-used. Why not simplify and unify naming for more similarity with the next clause?
```suggestion
    Debt = RoundedDelayMs - DelayMs,
    DebtInUnits = convert_time_unit(Debt, millisecond, TimeUnit),
```